sys/stdio_semihosting/stdio_semihosting.c

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (C) 2020 Koen Zandberg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     sys
+ * @{
+ *
+ * @file
+ * @brief       STDIO over ARM Semihosting implementation
+ *
+ * @author      Koen Zandberg <koen@bergzand.net>
+ *
+ * @}
+ */
+
+#include <errno.h>
+#include <string.h>
+
+#include "stdio_semihosting.h"
+#include "xtimer.h"
+
+#if MODULE_VFS
+#include "vfs.h"
+#endif
+
+/**
+ * @brief ARM Semihosting STDIN file descriptor
+ */
+#define STDIO_SEMIHOSTING_F_STDIN       (1)
+
+/**
+ * @brief ARM Semihosting STDOUT file descriptor
+ */
+#define STDIO_SEMIHOSTING_F_STDOUT      (1)
+
+/**
+ * @name ARM Semihosting commands
+ *
+ * Extend when required
+ * @{
+ */
+#define STDIO_SEMIHOSTING_SYS_WRITE     (0x05) /**< Write command */
+#define STDIO_SEMIHOSTING_SYS_READ      (0x06) /**< Read command  */
+/** @} */
+
+static bool _semihosting_connected(void) {
+#ifdef CoreDebug_DHCSR_C_DEBUGEN_Msk
+    /* Best effort attempt to detect if a debug session is active */
+    return CoreDebug->DHCSR & CoreDebug_DHCSR_C_DEBUGEN_Msk;
+#else
+    return true;
+#endif
+}
+
+static uint32_t _semihosting_raw(int cmd, uint32_t *args)
+{
+    uint32_t result = 0;
+    /* Moves cmd and args to r0 and r1. Then triggers a breakpoint.
+     * Finally moves the results stored in r0 to result
+     */
+    __asm__(
+        "mov r0, %[cmd] \n"
```

Comment:
Should we specify `.thumb_func` since that is the `thumb_func` since that is the T32 trap?

Reply:
I'm not familiar enough with this to answer this. Maybe @kaspar030 knows if this makes sense

Reply:
I think `.thumb_func` is used in front of a function label so the assembler creates the symbol with the thumb bit set. It shouldn't be needed here. Does it change the generated assembly (Thumb vs ARM) when present?

Reply:
Doesn't change anything, so I guess its not needed.